### PR TITLE
Remove ellipsis from theme

### LINKED ISSUE #1744: Eliminate ... in themes()
Instead using multiple argument names, and documenting them with `@param`

### PR DESCRIPTION
**Pull request by thomasp85**
Fixes #1744

### CODE REVIEW

File: R/theme.r [hunk 1]
```diff
@@ -385,8 +364,30 @@ print.theme <- function(x, ...) utils::str(x)
...
+                  panel.grid.major, panel.grid.minor, panel.grid.major.x,
+                  panel.grid.major.y, panel.grid.minor.x, panel.grid.minor.y,
+                  panel.ontop, plot.background, plot.title, plot.subtitle,
+                  plot.caption, plot.margin, strip.background, strip.text,
+                  strip.text.x, strip.text.y, strip.switch.pad.grid,
+                  strip.switch.pad.wrap, ..., complete = FALSE, validate = TRUE) {
+  elements <- as.list(environment())
+  elements$complete <- NULL
+  elements$validate <- NULL
+  elements <- elements[lapply(elements, class) != "name"]
```
Comment (thomasp85) [inline R/theme.r]: I would love some feedback on whether this is the best way to capture all non-missing arguments as evaluated values...?

File: R/theme.r [hunk 2]
```diff
@@ -80,166 +80,145 @@ print.theme <- function(x, ...) utils::str(x)
...
-#' @param ... a list of element name, element pairings that modify the
-#'   existing theme.
-#' @param complete set this to TRUE if this is a complete theme, such as
-#'   the one returned \code{by theme_grey()}. Complete themes behave
-#'   differently when added to a ggplot object.
+#' @param line all line elements (\code{element_line})
+#' @param rect all rectangular elements (\code{element_rect})
+#' @param text all text elements (\code{element_text})
+#' @param title all title elements: plot, axes, legends (\code{element_text};
+#' inherits from \code{text})
```
Comment (hadley) [inline R/theme.r]: Can you please indent the second and subsequent lines?

File: R/theme.r [hunk 3]
```diff
@@ -385,8 +364,30 @@ print.theme <- function(x, ...) utils::str(x)
...
+                  panel.grid.major, panel.grid.minor, panel.grid.major.x,
+                  panel.grid.major.y, panel.grid.minor.x, panel.grid.minor.y,
+                  panel.ontop, plot.background, plot.title, plot.subtitle,
+                  plot.caption, plot.margin, strip.background, strip.text,
+                  strip.text.x, strip.text.y, strip.switch.pad.grid,
+                  strip.switch.pad.wrap, ..., complete = FALSE, validate = TRUE) {
+  elements <- c(as.list(environment()), list(...))
+  elements$complete <- NULL
+  elements$validate <- NULL
+  elements <- elements[lapply(elements, class) != "name"]
```
Comment (hadley) [inline R/theme.r]: Why is this necessary? Regardless, it would be cleaner to use `vapply()` + `is.name()`

File: R/theme.r [hunk 4]
```diff
@@ -385,8 +364,30 @@ print.theme <- function(x, ...) utils::str(x)
...
+                  legend.box.margin, legend.box.background, legend.box.spacing,
+                  panel.background, panel.border, panel.spacing,
+                  panel.spacing.x, panel.spacing.y, panel.grid,
+                  panel.grid.major, panel.grid.minor, panel.grid.major.x,
+                  panel.grid.major.y, panel.grid.minor.x, panel.grid.minor.y,
+                  panel.ontop, plot.background, plot.title, plot.subtitle,
+                  plot.caption, plot.margin, strip.background, strip.text,
+                  strip.text.x, strip.text.y, strip.switch.pad.grid,
+                  strip.switch.pad.wrap, ..., complete = FALSE, validate = TRUE) {
+  elements <- c(as.list(environment()), list(...))
```
Comment (hadley) [inline R/theme.r]: Hmmmm, I don't like this but I don't see a better way except creating the massive named list by hand (which seems likely to be error prone).

File: R/theme.r (hunk 3 above)
Comment (thomasp85) [inline R/theme.r]: If we don't remove `name` elements the theme validation will complain

Comment (hadley) [inline R/theme.r]: But why? Where do those name elements come from?

Comment (thomasp85) [inline R/theme.r]: these are missing arguments - they exist in the environment and is added to the list as name objects

**Comment by hadley** [comment]
@wch I have vague memory you thought about this problem (capturing all arguments) a bit

File: R/theme.r [hunk 5]
```diff
@@ -385,8 +365,30 @@ print.theme <- function(x, ...) utils::str(x)
...
+  elements$complete <- NULL
+  elements$validate <- NULL
+  elements <- elements[!vapply(elements, is.name, logical(1))]
```
Comment (hadley) [inline R/theme.r]: The absolutely correct check is:

```R
elements <- elements[!vapply(elements, identical, quote(expr = ), FUN.VALUE = logical(1))]
```

File: R/theme.r (hunk 4 above)
Comment (hadley) [inline R/theme.r]: I would really prefer if this was extracted into a separate function and gained a couple of safe guards:

```R
find_args <- function(...) {
  env <- parent.frame()
  args <- names(formals(sys.function(sys.parent(1))))
  
  vals <- c(as.list(env)[args], list(...))
  vals <- vals[!vapply(vals, identical, quote(expr = ), FUN.VALUE = logical(1))]
  vals
}

f <- function(a = 1, b = 2, c = 3, d, e, abcdefasdfgasdfasdfasd) {
   find_args()
}
f(10, abc = 10)
```

Comment (hadley) [inline R/theme.r]: This is slightly more elegant:

```R
find_args <- function(...) {
  env <- parent.frame()
  args <- names(formals(sys.function(sys.parent(1))))
  
  vals <- mget(args, envir = env)
  vals <- vals[!vapply(vals, identical, quote(expr = ), FUN.VALUE = logical(1))]
  
  c(vals, list(...))
}
```

Comment (hadley) [inline R/theme.r]:
```R
is_missing_arg <- function(x) identical(x. quote(expr = )))
```

Comment (hadley) [inline R/theme.r]: And even better:

```R
find_args <- function(...) {
  env <- parent.frame()
  args <- names(formals(sys.function(sys.parent(1))))

  vals <- mget(args, envir = env)
  vals <- vals[!vapply(vals, identical, quote(expr = ), FUN.VALUE = logical(1))]

  modifyList(vals, list(...))
}
```

File: R/utilities.r [hunk 6]
```diff
@@ -281,6 +281,22 @@ dispatch_args <- function(f, ...) {
   f
 }
 
+is_missing_arg <- function(x) identical(x, quote(expr = ))
+# Get all arguments in a function as a list. Will fail if an ellipsis argument
+# named .ignore
+# @param ... passed on in case enclosing function uses ellipsis in argument list
+# @param .ignore remove these arguments from the list
+find_args <- function(..., .ignore = NULL) {
```
Comment (hadley) [inline R/utilities.r]: Why not drop `.ignore` and just use (e.g.) `complete = NULL, validate = NULL)`

File: NEWS.md [hunk 7]
```diff
@@ -56,6 +56,8 @@
 * Themes are more homogeneous visually, and match `theme_grey` better.
   (@jiho, #1679)
 
+* The `theme()` constructor now has named arguments rather than ellipsis.
```
Comment (hadley) [inline NEWS.md]: "This should make autocomplete substantially more useful"

File: R/utilities.r (hunk 6 above)
Comment (thomasp85) [inline R/utilities.r]: Because `validate_theme` will throw an error if they are included in the list

Comment (hadley) [inline R/utilities.r]:
```R
find_args(..., complete = NULL, validate = NULL)
```

Comment (hadley) [inline R/utilities.r]: Then `modifyList` will remove them

Comment (thomasp85) [inline R/utilities.r]: Ah - didn't think of that

File: R/utilities.r [hunk 8]
```diff
@@ -281,6 +281,21 @@ dispatch_args <- function(f, ...) {
   f
 }
 
+is_missing_arg <- function(x) identical(x, quote(expr = ))
+# Get all arguments in a function as a list. Will fail if an ellipsis argument
+# named .ignore
+# @param ... passed on in case enclosing function uses ellipsis in argument list
+# @param .ignore remove these arguments from the list
```
Comment (hadley) [inline R/utilities.r]: This needs to be removed now, other LGTM
